pkg/virtualization/core/gpu/gpu.go

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package gpu
+
+import (
+	"log"
+
+	"github.com/microsoft/wmi/pkg/base/host"
+	"github.com/microsoft/wmi/pkg/base/instance"
+	"github.com/microsoft/wmi/pkg/base/query"
+	"github.com/microsoft/wmi/pkg/constant"
+)
+
+// GetPartitionableGpuCollection gets all host partitionable GPUs
+func GetPartitionableGpuCollection(whost *host.WmiHost) (partitionablegpucollection PartitionableGpuCollection, err error) {
+	query := query.NewWmiQuery("Msvm_PartitionableGpu")
+	rasdcollection, err := instance.GetWmiInstancesFromHost(whost, string(constant.Virtualization), query)
+	if err != nil {
+		log.Printf("[WMI] Error getting Msvm_PartitionableGpu instances - Error details [%+v]\n", err)
+		return
+	}
+	defer rasdcollection.Close()
```

Comment:
This seems to be a problem. I dont see. Can you refer to other places how this is done.

If we close this, the returned collection is not valid

Reply:
It's being closed in most places where GetWmiInstancesFromHost is used.

I believe since we create a new collection (partitionablegpucollection) based on the WmiInstances (rasdcollection), it doesn't matter whether the original collection of WmiInstances is closed.

Reply:
New collection doesnt clone this object. Can you pass me the other places where this is done like this

Reply:
You shouldnt close it, if you return the wmiinstances.

You can close it if you plan to process it and you dont need them. This has to be fixed in other places, prob. from your previous PR

Reply:
You're right, I see lot of places where it gets closed only if err != nil. Do the callers of this method will have to close this collection, then?

Other places where this happens (I wonder how it hasn't caused any issues though):
- https://github.com/microsoft/wmi/blob/edf3c9799294458bbd5a47dcf6ded9014cf5fb6c/pkg/hardware/host/pnpentity.go#L18-L41
- https://github.com/microsoft/wmi/blob/edf3c9799294458bbd5a47dcf6ded9014cf5fb6c/pkg/hardware/host/virtualmachine.go#L17-L34

I can fix it in the above two places as well.

There were some places where it gets closed, but some processing happens, so as you said that's why it would be fine.

Reply:
There is no Garbage collection. We have to close it